### PR TITLE
Improve precision recall metric issue 2571

### LINKED ISSUE #2571: Add new options to Precision and Recall metrics
## 🚀 Feature

I suggest to add new options regarding `average` parameter of `Precision` and `Recall` metrics.
Currently they have two options which are `average==False` and `average==True`. I propose to change `average` options to the following:

- `False`: return per class/label metric for multiclass/label input and simple precision for binary input (default option). This is equivalent with `average=None` of Scikit-learn for multiclass/label input and `average=binary` for binary input.
- `micro`: for multilabel, computes metric using all samples of all labels altogether(equivalent with `micro` of Scikit-learn). Incompatible with binary and multiclass inputs, as it would be equivalent with `Accuracy` and user can use that metric.
- `weighted`: It computes metric for every class/label then returns weighted average of them using number of samples in each class/label. Recall does not have this option because weighted recall is equivalent with accuracy. Equivalent with `weighted` of Scikit-Learn.
- `samples`: just for multilabel input. It computes metric for every sample then returns their average. Equivalent with `samples` of Scikit-Learn.

I did not put `macro` option because user can instantiate metric with `average=False` then call `mean` on it to become macro metric.

### PR DESCRIPTION
**Pull request by sadra-barikbin**
Fixes #2571 

Check list:

- [ ] New tests are added (if a new feature is added)
- [ ] New doc strings: description and/or example code are in RST format
- [x] Documentation is updated (if required)

### CODE REVIEW

File: ignite/metrics/recall.py [hunk 1]
```diff
@@ -171,6 +167,9 @@ def __init__(
         is_multilabel: bool = False,
         device: Union[str, torch.device] = torch.device("cpu"),
     ):
+
+        if average not in [False, "micro", "samples"]:
```
Comment (vfdev-5) [inline ignite/metrics/recall.py:171]: Why Recall can't be weighted ?

File: ignite/metrics/precision.py [hunk 2]
```diff
@@ -93,139 +126,137 @@ class Precision(_BasePrecisionRecall):
...
     def __init__(
         self,
         output_transform: Callable = lambda x: x,
-        average: bool = False,
+        average: Union[bool, str] = False,
         is_multilabel: bool = False,
         device: Union[str, torch.device] = torch.device("cpu"),
     ):
+
+        if average not in [False, "micro", "weighted", "samples"]:
```
Comment (vfdev-5) [inline ignite/metrics/precision.py:258]: We should still support average True and translate it to appropriate str value

**Review by vfdev-5 — COMMENTED** [review]
@sadra-barikbin thanks a lot for the PR ! 
I left few comments to clear out the API. I haven't checked yet the details but the feature looks cool !
Please check also docs build failure.

File: ignite/metrics/precision.py [hunk 3]
```diff
@@ -15,7 +15,7 @@ class _BasePrecisionRecall(_BaseClassification):
     def __init__(
         self,
         output_transform: Callable = lambda x: x,
-        average: bool = False,
+        average: Union[bool, str] = False,
```
[the remark this answers is not in the document]
Comment (sadra-barikbin) [reply ignite/metrics/precision.py:18]: For the first point, I've done it in `Precision` and `Recall` themselves because their set of `average` options are not the same. I didn't get the second point. You mean if we got `False`, set `average` to `'false'` to be like other options?

File: ignite/metrics/recall.py (hunk 1 above)
Comment (sadra-barikbin) [reply ignite/metrics/recall.py:178]: Weighted recall is equivalent with accuracy. For example, in binary case we have:
image

in which 
image

File: ignite/metrics/precision.py (hunk 2 above)
Comment (sadra-barikbin) [reply ignite/metrics/precision.py:266]: Which option do you think is better to translate to? For precision, I suggest `weighted`, as it's `macro` plus considering classes/labels imbalance.

File: ignite/metrics/recall.py (hunk 1 above)
Comment (sadra-barikbin) [reply ignite/metrics/recall.py:178]: This is stated [here](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.recall_score.html) as well.

Comment (vfdev-5) [reply ignite/metrics/recall.py:178]: I think it is not a problem that Weighted Recall == Accuracy. IMO we could support it as scikit-learn does.

File: ignite/metrics/classification_report.py [hunk 4]
```diff
@@ -4,7 +4,6 @@
 import torch
```
Comment (vfdev-5) [inline ignite/metrics/classification_report.py:4]: Let's keep classification report updates to another PR

File: ignite/metrics/precision.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (vfdev-5) [reply ignite/metrics/precision.py:18]: You are right we are missing `None` value as scikit-learn which would match previous `average=False`. We should also deprecate `is_multilabel` as it becomes unused. What do you think ?

> I've done it in Precision and Recall themselves because their set of average options are not the same.

Can you detail a bit more this point ? I haven't checked but IMO they should be same.

File: ignite/metrics/precision.py (hunk 2 above)
Comment (vfdev-5) [reply ignite/metrics/precision.py:266]: Translation should be without any BC-breaking changes, so I'd say `macro`.

File: ignite/metrics/recall.py (hunk 1 above)
Comment (sadra-barikbin) [reply ignite/metrics/recall.py:178]: So, supported options for `Precision` and `Recall` would be equal.

File: ignite/metrics/precision.py (hunk 2 above)
Comment (sadra-barikbin) [reply ignite/metrics/precision.py:266]: Got it

File: ignite/metrics/classification_report.py (hunk 4 above)
Comment (sadra-barikbin) [reply ignite/metrics/classification_report.py:4]: Its tests woulf fail, then. Shall I mark them as `xfail`?

File: ignite/metrics/precision.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sadra-barikbin) [reply ignite/metrics/precision.py:18]: Yes, we could remove `multilabel`. It could be infered from data.
`Recall` has not the `weighted` option, but regarding the next comment, I'll add it so they become the same.

File: ignite/metrics/classification_report.py (hunk 4 above)
Comment (vfdev-5) [reply ignite/metrics/classification_report.py:4]: This is curious why the tests are failing then. As right now tests are passing and we introducing new features trying to keep BC, I would expect that tests won't break...

File: ignite/metrics/precision.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (vfdev-5) [reply ignite/metrics/precision.py:18]: I was thinking that `average="samples"` is equivalent to `is_multilabel=True, average=True`. Can we use `average="samples"` for binary case ?

[the remark this answers is not in the document]
Comment (sadra-barikbin) [reply ignite/metrics/precision.py:18]: `average='samples'` for binary case is equivalent with the term below which is not a known metric:
image
According to Scikit, `average='samples'` is "only meaningful for multilabel classification".
Anyway, without `is_multilabel`, we cannot differentiate binary data from multilabel data because we do not have a differentiator assumption on their dimensions. Scikit can, because it assumes binary data is 1-dimensional and multilabel data is 2-dimensional.

File: ignite/metrics/classification_report.py (hunk 4 above)
Comment (sadra-barikbin) [reply ignite/metrics/classification_report.py:4]: You're right. They had been marked as `xfail` before.

File: tests/ignite/metrics/test_precision.py [hunk 5]
```diff
@@ -345,58 +357,40 @@ def get_test_cases():
...
-        assert pr._updated is False
+    with pytest.raises(ValueError):
+        pr.update((y_pred, y))
 
-    _test(average=True)
-    _test(average=False)
+    assert pr._updated is False
 
 
 def _test_distrib_integration_multiclass(device):
```
Comment (sadra-barikbin) [inline tests/ignite/metrics/test_precision.py:396]: Is there any reason that this test has not been parametrized?

**Comment by sadra-barikbin** [comment]
Scikit-Learn has another data type called `multilabel-multioutput` which is like multilabel but its labels are not binary. It has yet to be implemented. We can do it if you will.

File: tests/ignite/metrics/test_precision.py (hunk 5 above)
Comment (vfdev-5) [reply tests/ignite/metrics/test_precision.py:397]: This test and other similar ones are used in test_distrib_hvd, test_distrib_gloo_cpu_or_gpu, test_distrib_nccl_gpu tests. IMO, it is not obvious how to properly parametrize them. However, there was a task to rewrite distributed tests such that we do not need to write all: `test_distrib_hvd, test_distrib_gloo_cpu_or_gpu, test_distrib_nccl_gpu` but only `_test_distrib_integration_multiclass` and they could be executed with hvd, gloo, nccl confs...

**Comment by vfdev-5** [comment]
> Scikit-Learn has another data type called multilabel-multioutput which is like multilabel but its labels are not binary. It has yet to be implemented. We can do it if you will.

Let's make it in a separate PR.

File: ignite/metrics/precision.py [hunk 6]
```diff
@@ -15,28 +15,39 @@ class _BasePrecisionRecall(_BaseClassification):
...
         is_multilabel: bool = False,
         device: Union[str, torch.device] = torch.device("cpu"),
     ):
 
+        if type(average) != bool and average not in ["micro", "weighted", "samples"]:
```
Comment (vfdev-5) [inline ignite/metrics/precision.py:23]: How about doing this like that ?
```suggestion
        if not (isinstance(average, bool) or average in ["micro", "weighted", "samples"]):
```

File: ignite/metrics/precision.py (hunk 2 above)
Comment (vfdev-5) [reply ignite/metrics/precision.py]: So, we need to add "macro" option

File: ignite/metrics/precision.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (vfdev-5) [reply ignite/metrics/precision.py:18]: I may be misunderstand your point, but I was thinking about new API like that:
```python
pr = Precision(average="samples")  # old: Precision(is_multilabel=True, average=True)
pr.update(multilabel_data)

pr = Precision()  # old: Precision()
pr.update(binary_data)

pr = Precision(average="samples")  # old: Precision(is_multilabel=True, average=True)
pr.update(binary_data)  # try to interpret binary_data as multilabel
```

> Anyway, without is_multilabel, we cannot differentiate binary data from multilabel data because we do not have a differentiator assumption on their dimensions. Scikit can, because it assumes binary data is 1-dimensional and multilabel data is 2-dimensional.

By the way, scikit learn has "binary" option as well. If we stuck with that, we can add it as well...

[the remark this answers is not in the document]
Comment (sadra-barikbin) [reply ignite/metrics/precision.py:18]: Do you still think we can remove `is_multilabel`?
About `average='binary'`, Scikit uses it only for binary data. When average is binary, it returns the precision for class `pos_label` which is another parameter. So if `pos_label` is 1 (default), the simple precision is returned, and if the `pos_label` is 0, `TN/(TN+FN)` (precision for class 0) is returned.
I think we don't need this option. If user wants precision for class 0, he/she can simply do `1-y` and `1-y_pred` in the `output_transform`.
In our current implementation, when average is `True`, we return the simple precision (precision for class 1) for binary data.

File: ignite/metrics/precision.py (hunk 2 above)
Comment (sadra-barikbin) [reply ignite/metrics/precision.py]: Didn't we assign `average=True` to macro?

File: ignite/metrics/precision.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (vfdev-5) [reply ignite/metrics/precision.py:18]: > Do you still think we can remove is_multilabel?

We do not remove it but just deprecate for removing later. For me if average can take str values, e.g. "samples", `is_multilabel` is redundant and can be deprecated. What do you think ?

[the remark this answers is not in the document]
Comment (sadra-barikbin) [reply ignite/metrics/precision.py:18]: If we depracate `is_multilabel`, how could we differentiate binary from multilabel in the case average is anything but 'samples'?

[the remark this answers is not in the document]
Comment (vfdev-5) [reply ignite/metrics/precision.py:18]: If average is not 'samples' then input data is not multilabel: binary or multiclass.

File: ignite/metrics/precision.py [hunk 7]
```diff
@@ -15,28 +15,39 @@ class _BasePrecisionRecall(_BaseClassification):
     def __init__(
         self,
         output_transform: Callable = lambda x: x,
-        average: bool = False,
+        average: Union[bool, str] = False,
         is_multilabel: bool = False,
         device: Union[str, torch.device] = torch.device("cpu"),
     ):
 
+        if not (isinstance(average, bool) or average in ["micro", "weighted", "samples"]):
```
Comment (vfdev-5) [inline ignite/metrics/precision.py:23]: When I was asking to add "macro", I wanted to add it to acceptable values `average in ["micro", "macro", "weighted", "samples"]`. I agree that `average=True` is the same as `average="macro"`. But it is not bad. We are gradually deprecating previous api to the new one with average as str similar to scikit-learn.
By the way, we have to add `average=None` as well which maps to `average=False`. 


What do you think about trying to match Scikit-learn API for average arg ?

File: ignite/metrics/precision.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sadra-barikbin) [reply ignite/metrics/precision.py:18]: Other options of average are also compatible with multilabel data, so if average is not 'samples', intput data can still be multilabel.

[the remark this answers is not in the document]
Comment (vfdev-5) [reply ignite/metrics/precision.py:18]: > Other options of average are also compatible with multilabel data, so if average is not 'samples', intput data can still be multilabel.

Can you give a concrete example ? Even if it is so, is output still meaningful and can be compared to scikit learn ?

File: ignite/metrics/precision.py (hunk 7 above)
Comment (sadra-barikbin) [reply ignite/metrics/precision.py:23]: Useful for users familiar with Scikit. If so, shall I add `pos_label` as well? (for `average='binary'`) As I said above, `pos_label` is another parameter which is used to determine precision/recall of which class be returned.

Comment (vfdev-5) [reply ignite/metrics/precision.py:23]: Yes, it can be useful, but let's do that in another PR, not here

**Comment by sadra-barikbin** [comment]
The formulas I just added to docstrings are for multi label as well. As an example, suppose we have exam result of some students and  exam questions are YES/NO questions. Exam results is multilabel data. If we measure precision considering questions altogether, it would be micro precision. If we calculate  precision for each question then measure (weighted) average of them, it would be (weighted) macro precision. 

Yes. I adopted weighted, samples and micro from Scikit, and basically used it in tests as well.

File: ignite/metrics/precision.py (hunk 7 above)
Comment (sadra-barikbin) [reply ignite/metrics/precision.py:23]: On the other hand , it decreases our api's elegance.

File: ignite/metrics/precision.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sadra-barikbin) [reply ignite/metrics/precision.py:18]: The formulas I just added to docstrings are for multi label as well. As an example, suppose we have exam result of some students and  exam questions are YES/NO questions. Exam results is multilabel data. If we measure precision considering questions altogether, it would be micro precision. If we calculate  precision for each question then measure (weighted) average of them, it would be (weighted) macro precision.

Yes. I adopted weighted, samples and micro from Scikit, and basically used it in tests as well.

File: ignite/metrics/precision.py [hunk 8]
```diff
@@ -15,28 +15,47 @@ class _BasePrecisionRecall(_BaseClassification):
...
+        if self._type in ["binary", "multiclass"] and self._average == "samples":
+            raise ValueError("Argument average='samples' is incompatible with binary and multiclass input data.")
+
     @reinit__is_reduced
     def reset(self) -> None:
-        self._true_positives = 0  # type: Union[int, torch.Tensor]
-        self._positives = 0  # type: Union[int, torch.Tensor]
-        self._updated = False
+        if self._average == "samples":
+            self._sum_samples_metric = 0  # type: Union[int, torch.Tensor]
```
Comment (vfdev-5) [inline ignite/metrics/precision.py:50]: @sadra-barikbin you introduce here new accumulators: `_sum_samples_metric`, `_samples_cnt`, `_actual_positives` in addition to existing `_true_positives`, `_positives`. Can we just keep 3 instead, we can rename them to something abstract if names like `_true_positives` and `_positives` are inappropriate.

File: ignite/metrics/precision.py [hunk 9]
```diff
@@ -250,31 +322,32 @@ def update(self, output: Sequence[torch.Tensor]) -> None:
             indices = torch.argmax(y_pred, dim=1).view(-1)
             y_pred = to_onehot(indices, num_classes=num_classes)
         elif self._type == "multilabel":
-            # if y, y_pred shape is (N, C, ...) -> (C, N x ...)
-            num_classes = y_pred.size(1)
-            y_pred = torch.transpose(y_pred, 1, 0).reshape(num_classes, -1)
-            y = torch.transpose(y, 1, 0).reshape(num_classes, -1)
+
```
Comment (vfdev-5) [inline ignite/metrics/precision.py:325]: Can you keep and update the comment you removed in this line

File: ignite/metrics/fbeta.py [hunk 10]
```diff
@@ -56,8 +56,8 @@ def Fbeta(
 
         .. testcode:: 1
 
-            P = Precision(average=False)
-            R = Recall(average=False)
+            P = Precision(average=False)[1] # `[1]` is to select the value for class 1 that is TP/(TP+FP)
+            R = Recall(average=False)[1] # `[1]` is to select the value for class 1 that is TP/(TP+FN)
```
Comment (vfdev-5) [inline ignite/metrics/fbeta.py:60]: @sadra-barikbin What we changed exactly in this PR (please provide the line of code) such that we have to update this example. I'm trying to understand if we are doing BC-breaking update or not

File: ignite/metrics/precision.py [hunk 11]
```diff
@@ -46,24 +65,33 @@ def compute(self) -> Union[torch.Tensor, float]:
...
-        if self._average:
-            return cast(torch.Tensor, result).mean().item()
+        result = self._true_positives / (self._positives + self.eps)
+        if self._average == "weighted":
+            denominator = self._actual_positives.sum() + self.eps  # type: ignore
+            return ((result @ self._actual_positives) / denominator).item()  # type: ignore
+        elif self._average == "micro":
+            return result.item()  # type: ignore
+        elif self._average == "macro":
+            return result.mean().item()  # type: ignore
```
Comment (vfdev-5) [inline ignite/metrics/precision.py:88]: Why we have to `type: ignore` ?

File: ignite/metrics/precision.py [hunk 12]
```diff
@@ -93,153 +179,139 @@ class Precision(_BasePrecisionRecall):
...
 
         .. testcode:: 1
 
-            metric = Precision(average=False)
+            metric = Precision()
+            weighted_metric = Precision(average='weighted')
             metric.attach(default_evaluator, "precision")
-            y_true = torch.Tensor([1, 0, 1, 1, 0, 1])
+            weighted_metric.attach(default_evaluator, "weighted precision")
+            y_true = torch.Tensor([1, 0, 1, 1, 0, 1]).long()
```
Comment (vfdev-5) [inline ignite/metrics/precision.py:191]: @sadra-barikbin please can you check and replace everywhere in the docstrings (Precision, Recall) that we create tensors like `torch.tensor` and not `torch.Tensor`. Thanks !

File: ignite/metrics/precision.py [hunk 13]
```diff
@@ -250,31 +322,32 @@ def update(self, output: Sequence[torch.Tensor]) -> None:
...
             y_pred = to_onehot(indices, num_classes=num_classes)
         elif self._type == "multilabel":
-            # if y, y_pred shape is (N, C, ...) -> (C, N x ...)
-            num_classes = y_pred.size(1)
-            y_pred = torch.transpose(y_pred, 1, 0).reshape(num_classes, -1)
-            y = torch.transpose(y, 1, 0).reshape(num_classes, -1)
+
+            num_labels = y_pred.size(1)
+            y_pred = torch.transpose(y_pred, 1, -1).reshape(-1, num_labels)
+            y = torch.transpose(y, 1, -1).reshape(-1, num_labels)
```
Comment (vfdev-5) [inline ignite/metrics/precision.py:328]: Why it is better to make it (N * ..., C) instead of (C, N * ...) ?

File: ignite/metrics/recall.py [hunk 14]
```diff
@@ -40,151 +101,127 @@ class Recall(_BasePrecisionRecall):
...
     def update(self, output: Sequence[torch.Tensor]) -> None:
         self._check_shape(output)
         self._check_type(output)
         y_pred, y = output[0].detach(), output[1].detach()
 
         if self._type == "binary":
-            y_pred = y_pred.view(-1)
-            y = y.view(-1)
+
+            y = to_onehot(y.view(-1), num_classes=2)
```
Comment (vfdev-5) [inline ignite/metrics/recall.py:221]: Do you think we can refactor this code to avoid code duplication ?

File: tests/ignite/metrics/test_metrics_lambda.py [hunk 15]
```diff
@@ -149,7 +149,7 @@ def test_metrics_lambda_update():
     recall = Recall(average=False)
 
     def Fbeta(r, p, beta):
-        return torch.mean((1 + beta ** 2) * p * r / (beta ** 2 * p + r)).item()
+        return (1 + beta ** 2) * p * r / (beta ** 2 * p + r)
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_metrics_lambda.py:152]: Why this change ?

File: ignite/metrics/precision.py [hunk 16]
```diff
@@ -93,153 +179,139 @@ class Precision(_BasePrecisionRecall):
...
     def update(self, output: Sequence[torch.Tensor]) -> None:
         self._check_shape(output)
         self._check_type(output)
         y_pred, y = output[0].detach(), output[1].detach()
 
         if self._type == "binary":
-            y_pred = y_pred.view(-1)
-            y = y.view(-1)
+
+            y = to_onehot(y.view(-1), num_classes=2)
```
Comment (vfdev-5) [inline ignite/metrics/precision.py:311]: OK, I see, you pass binary case into multiclass (2-classes) case. I think we have to keep computed result for the single class (=1) instead of both. Our public API and usage should not change and this PR shouldn't be BC-breaking for this case.

File: ignite/metrics/precision.py (hunk 8 above)
Comment (sadra-barikbin) [reply ignite/metrics/precision.py:50]: Yes. For example `nominator`, `denominator` and `weight`. It hurts readability, though. Not only the type but also the semantic of `_sum_samples_metric` is different with  that of `_true_positives` or `_positives`. The former is a float value representing the summation of computed metric over samples and the latter is the number of samples.

File: tests/ignite/metrics/test_metrics_lambda.py (hunk 15 above)
Comment (sadra-barikbin) [reply tests/ignite/metrics/test_metrics_lambda.py:152]: As we agreed, `average=False` results in precision and recall becoming an array with length equal to number of classes. Had I not changed  that line, it would have returned the average of Fbeta for class zero and Fbeta for class one. Now it returns Fbeta array and I pick Fbeta for class one in line 215

File: ignite/metrics/fbeta.py (hunk 10 above)
Comment (sadra-barikbin) [reply ignite/metrics/fbeta.py:60]: https://github.com/sadra-barikbin/ignite/blob/214088b94c21471e4e287cefc411bc2f45bab3bb/ignite/metrics/precision.py#L311-L312

Comment (sadra-barikbin) [reply ignite/metrics/fbeta.py:60]: Alternatively I can do this:
```python
P = Precision(average=False)
R = Recall(average=False)
metric = Fbeta(beta=1.0, precision=P, recall=R, average=False)[1]
```

Comment (sadra-barikbin) [reply ignite/metrics/fbeta.py:60]: This change is because beforehand, Precision or recall with `average=False` returned a single float on binary input, but now they return an array with length 2.

File: ignite/metrics/precision.py (hunk 11 above)
Comment (sadra-barikbin) [reply ignite/metrics/precision.py:88]: Because we have
```python
result = self._true_positives / (self._positives + self.eps)
```
and `_true_positives` and `_positives` are of typehint `Union[int, torch.Tensor]` . So `result`'s type would implicitly be `Union[float, torch.Tensor]`. In line 88 we call `mean()` on `result`. On the other hand each feasible type of `result` should have `mean` method, but float has not. So MyPy raises error if I do not add the hint

File: ignite/metrics/precision.py (hunk 13 above)
Comment (sadra-barikbin) [reply ignite/metrics/precision.py:328]: Because in binary and multilabel case, y and y_pred finally become of size N*C. I did the same thing to multilabel to reuse the logic in following lines.

File: ignite/metrics/precision.py (hunk 8 above)
Comment (sadra-barikbin) [reply ignite/metrics/precision.py]: I added comment to `reset` and `compute` functions which are also common between precision and recall. Is it ok?

File: tests/ignite/metrics/test_metrics_lambda.py (hunk 15 above)
Comment (vfdev-5) [reply tests/ignite/metrics/test_metrics_lambda.py:152]: > As we agreed, average=False results in precision and recall becoming an array with length equal to number of classes.

It was the same behaviour previously as well. That's why `torch.mean(...).item()` was applied to compute an average. I think we should keep tests without much changes to ensure BC.

File: tests/ignite/metrics/test_metrics_lambda.py [hunk 17]
```diff
@@ -195,24 +195,24 @@ def Fbeta(r, p, beta):
...
+    assert recall._nominator[1] == true_positives
+    assert recall._denominator[1] == positives
 
     """
     Test compute
     """
     F1.reset()
     F1.update((y_pred1, y1))
-    F1_metrics_lambda = F1.compute()
+    F1_metrics_lambda = F1.compute()[1]
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_metrics_lambda.py:215]: Following previous comment above. Let's revert this and above changes.

File: tests/ignite/metrics/test_metrics_lambda.py [hunk 18]
```diff
@@ -400,8 +400,8 @@ def data(y_pred, y):
         np_y = y.numpy().ravel()
         assert state.metrics[metric_name] == approx(compute_true_value_fn(np_y_pred, np_y))
 
-    precision_1 = Precision()
-    precision_2 = Precision()
+    precision_1 = Precision()[1]
```
Comment (vfdev-5) [inline tests/ignite/metrics/test_metrics_lambda.py:403]: Same here, I'd prefer to revert these changes as well

File: ignite/metrics/precision.py [hunk 19]
```diff
@@ -15,55 +15,111 @@ class _BasePrecisionRecall(_BaseClassification):
...
-                self._true_positives = cast(torch.Tensor, idist.all_gather(self._true_positives))
-                self._positives = cast(torch.Tensor, idist.all_gather(self._positives))
+            self._nominator = idist.all_reduce(self._nominator)  # type: ignore[assignment]
+            self._denominator = idist.all_reduce(self._denominator)  # type: ignore[assignment]
+            if self._average == "weighted":
+                self._weight = idist.all_reduce(self._weight)  # type: ignore[assignment]
             self._is_reduced = True  # type: bool
 
-        result = self._true_positives / (self._positives + self.eps)
+        fraction = self._nominator / (self._denominator + (self.eps if self._average != "samples" else 0))
```
[the remark this answers is not in the document]
Comment (sadra-barikbin) [reply ignite/metrics/precision.py:108]: You mean to use eps for average=samples as well?

**Review by vfdev-5 — APPROVED** [review]
Thanks a lot for this great PR @sadra-barikbin !
I'm checking it on GPU and if CI is green let's merge it.
In a follow-up PR, let's clean-up docstring about 
```diff
- torch.Tensor(...)
+ torch.tensor(...)
```
